stan/math/opencl/has_event_stack.hpp

```diff
@@ -0,0 +1,66 @@
+#ifndef STAN_MATH_OPENCL_HAS_EVENT_STACK_HPP
+#define STAN_MATH_OPENCL_HAS_EVENT_STACK_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <CL/cl.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+namespace opencl_kernels {
+
+/**
+ * Template check for if type has event stack
+ * @tparam Type to check.
+ * @return enum with false value if type does not have an event stack.
+ */
+template <typename T>
+struct has_event_stack {
```

Comment:
I'm not sure we need this template metaprogram at all - right now we only have `matrix_cl` with events and I'm having trouble imagining a future where we add more types that have events... but maybe that's just my lack of imagination? Are you thinking of some we might need in the future?

Reply:
Oh, maybe I'm dumb - you could do this for Eigen types with caches too?

Reply:
> Are you thinking of some we might need in the future?

Nope! We totally don't need these, but the templates were getting wicked gross so I pulled these out to be a bit more dry.

> Oh, maybe I'm dumb - you could do this for Eigen types with caches too?

Not totally clear. Maybe?